python/ql/lib/semmle/python/security/regexp/NfaUtils.qll

```diff
@@ -0,0 +1,1323 @@
+    isReDoSAttackable(t, pump, s) and
+    (
+      prefixMsg = "starting with '" + escape(PrefixConstruction::prefix(s)) + "' and " and
+      not PrefixConstruction::prefix(s) = ""
+      or
+      PrefixConstruction::prefix(s) = "" and prefixMsg = ""
+      or
+      not exists(PrefixConstruction::prefix(s)) and prefixMsg = ""
+    )
+  }
+
+  /**
+   * Gets the result of backslash-escaping newlines, carriage-returns and
+   * backslashes in `s`.
+   */
+  bindingset[s]
+  private string escape(string s) {
+    result =
+      s.replaceAll("\\", "\\\\")
+          .replaceAll("\n", "\\n")
+          .replaceAll("\r", "\\r")
+          .replaceAll("\t", "\\t")
+  }
+
+  /**
+   * Gets `str` with the last `i` characters moved to the front.
+   *
+   * We use this to adjust the pump string to match with the beginning of
+   * a RegExpTerm, so it doesn't start in the middle of a constant.
+   */
+  bindingset[str, i]
+  private string rotate(string str, int i) {
+    result = str.suffix(str.length() - i) + str.prefix(str.length() - i)
+  }
+}
+
+/**
+ * A module that describes a tree where each node has one or more associated characters, also known as a trie.
+ * The root node has no associated character.
+ * This module is a signature used in `Concretizer`.
+ */
+signature module CharTree {
+  /** A node in the tree. */
+  class CharNode;
+
+  /** Gets the previous node in the tree from `t`. */
+  CharNode getPrev(CharNode t);
+
+  /**
+   * Holds if `n` is at the end of a tree. I.e. a node that should have a result in the `Concretizer` module.
+   * Such a node can still have children.
+   */
+  predicate isARelevantEnd(CharNode n);
+
+  /** Gets a char associated with `t`. */
+  string getChar(CharNode t);
+}
+
+/**
+ * Implements an algorithm for computing all possible strings
+ * from following a tree of nodes (as described in `CharTree`).
+ *
+ * The string is build using one big concat, where all the chars are computed first.
+ * See `concretize`.
+ */
+module Concretizer<CharTree Impl> {
+  private class Node = Impl::CharNode;
+
+  private predicate getPrev = Impl::getPrev/1;
+
+  private predicate isARelevantEnd = Impl::isARelevantEnd/1;
+
+  private predicate getChar = Impl::getChar/1;
+
+  /** Holds if `n` is on a path from the root to a leaf, and is therefore relevant for the results in `concretize`. */
+  private predicate isRelevant(Node n) {
+    isARelevantEnd(n)
+    or
+    exists(Node prev | isRelevant(prev) | n = getPrev(prev))
+  }
+
+  /** Holds if `n` is a root with no predecessors. */
+  private predicate isRoot(Node n) { not exists(getPrev(n)) }
+
+  /** Gets the distance from a root to `n`. */
+  private int nodeDepth(Node n) {
+    result = 0 and isRoot(n)
+    or
+    isRelevant(n) and
+    exists(Node prev | result = nodeDepth(prev) + 1 | prev = getPrev(n))
+  }
+
+  /** Gets an ancestor of `end`, where `end` is a node that should have a result in `concretize`. */
+  private Node getANodeInLongChain(Node end) {
```

Comment:
Could this just be called `getAnAncestor`? And could it be written via `prev*`?

Reply:
Maybe it can, I don't remember if I wrote the recursion manually for a reason 🤷  

I'll do another evaluation to check.